diff -up openssl-0.9.8k/crypto/x509/x509_cmp.c.name-cmp openssl-0.9.8k/crypto/x509/x509_cmp.c
--- openssl-0.9.8k/crypto/x509/x509_cmp.c.name-cmp	2009-02-15 13:10:39.000000000 +0100
+++ openssl-0.9.8k/crypto/x509/x509_cmp.c	2009-03-25 20:04:41.000000000 +0100
@@ -282,15 +282,7 @@ int X509_NAME_cmp(const X509_NAME *a, co
 		nb=sk_X509_NAME_ENTRY_value(b->entries,i);
 		j=na->value->type-nb->value->type;
 		if (j)
-			{
-			nabit = ASN1_tag2bit(na->value->type);
-			nbbit = ASN1_tag2bit(nb->value->type);
-			if (!(nabit & STR_TYPE_CMP) ||
-				!(nbbit & STR_TYPE_CMP))
-				return j;
-			if (!asn1_string_memcmp(na->value, nb->value))
-				j = 0;
-			}
+			return j;
 		else if (na->value->type == V_ASN1_PRINTABLESTRING)
 			j=nocase_spacenorm_cmp(na->value, nb->value);
 		else if (na->value->type == V_ASN1_IA5STRING
