--- openssl-0.9.8b/crypto/engine/eng_all.c.padlock	2005-04-19 15:24:44.000000000 +0200
+++ openssl-0.9.8b/crypto/engine/eng_all.c	2006-06-21 19:57:29.000000000 +0200
@@ -69,6 +69,9 @@
 	ENGINE_load_openssl();
 #endif
 	ENGINE_load_dynamic();
+#if !defined(OPENSSL_NO_HW) && !defined(OPENSSL_NO_HW_PADLOCK)
+	ENGINE_load_padlock();
+#endif
 #ifndef OPENSSL_NO_STATIC_ENGINE
 #ifndef OPENSSL_NO_HW
 #ifndef OPENSSL_NO_HW_4758_CCA
@@ -95,9 +98,6 @@
 #ifndef OPENSSL_NO_HW_UBSEC
 	ENGINE_load_ubsec();
 #endif
-#ifndef OPENSSL_NO_HW_PADLOCK
-	ENGINE_load_padlock();
-#endif
 #endif
 #if defined(__OpenBSD__) || defined(__FreeBSD__)
 	ENGINE_load_cryptodev();
