Index: openssl/crypto/bn/asm/s390x.S
RCS File: /v/openssl/cvs/openssl/crypto/bn/asm/s390x.S,v
rcsdiff -q -kk '-r1.1.4.1' '-r1.1.4.2' -u '/v/openssl/cvs/openssl/crypto/bn/asm/s390x.S,v' 2>/dev/null
--- openssl/crypto/bn/asm/s390x.S 2010/09/10 14:55:24 1.1.4.1
+++ openssl/crypto/bn/asm/s390x.S 2010/11/22 21:57:07 1.1.4.2
@@ -131,7 +131,7 @@
 	brct	%r4,.Loop4_mul
 
 	la	%r10,1(%r10)		// see if len%4 is zero ...
-	brct	%r10,.Loop1_madd	// without touching condition code:-)
+	brct	%r10,.Loop1_mul		// without touching condition code:-)
 
 .Lend_mul:
 	alcgr	%r8,zero	// collect carry bit
